README
			      GNU Chess

Copyright (C) 1986 Stuart Cracraft
  (Copying permission notice at the end.)

GNU Chess is your program (as long as you follow the copyright and
licensing rules listed in the file COPYING). Your contribution to GNU
Chess will be retained and given to hundreds of other GNU Chess
devotees. Each improvement you make is important not just to you, but
also to all of us. As such, you must give us your changes. And in
fact, you are required to do so.

This program is one small step toward making generally available a
chess program with source to inspire current and future software
developers. This program is purely experimental. It is not a
strong program.

In the best of all possible worlds, not only would this program play
totally legal chess, but it would also play well. I leave this up to
the recipients of this document, fellow computer chess enthusiasts in
the world at large.

Move generator
--------------

The most important thing in any chess program is the speed of the
legal move generator. How fast can you generate lists of legal moves
for arbitrary chess positions? Research has shown that this is
directly correlated to how many moves ahead you can look and therefore
inextricably linked with the chess strength of the chess program.

This program has an adequate move generator. But it is nothing to
boast about. The move generator is in the module 'gen.c'.  An
alternate move generator is contained in the files tbl_*.c.  The
latter is a table-driven move generator in which all offsets normally
calculated upon execution are precalculated. It might work faster
than the non-table-driven version unless the memory accessing is
particularly slow. However, tests have shown that overall,
table-driven is inferior due to the increased complexity.  (On a Vax
8650, the GNU Chess move generator produces about 1200 move generations
per second.  While this is adequate, it is definitely not fast enough.)

If you can write a move generator that is super-fast, you are well on
your way to writing a skillful artificial chessplayer.  But if you
manage to do this, please let me know! I have spent many hours upon
move generators, and still don't have anything I consider speedy.

Opening Book
------------

GNU Chess has an opening book. It consists of hundreds of master games
from tournament play. The program converts these games into positions,
independent of moves, and then constructs a database such that any
position may be retrieved via a hash-address. 

There are currently two styles of syntax supported. The first is
in "bookin". The second is in bookin.xxx, where "xxx" refers to
the type of games in the collection. For example, the syntax of
bookin.tal, reduced algebraic, is the second supported syntax.
We do not support English descriptive syntax.

Transpositions are handled by GNU Chess because the positions are
dealt with independently of the series of moves by which they were
reached.  Any position which ever appears in the current game at ply-1
will be retrieved from the opening book along with a move made by the
a master, and more frequently an international master or grandmaster.
It does not matter when the position appears or what sequence of moves
has occurred before. If several suggested moves are available, the
program chooses from amongst these randomly, if -DRANDOM flag is set
within Makefile.  Without this flag, GNU Chess selects the first
available variation as it appears in physical order in the
human-readable book.

Now that the opening book C code has been written, our next
project is to type in all of the collected games of a great
master (such as Mikhail Tal or Robert Fischer). Then, GNU Chess
will simply play exactly like the great master in the opening,
assuming that the opponent obliges. You may wonder, but
what happens when the game has clearly gone out of book? What
to do? Just revert to the ordinary search? This is normally
the case. But with GNU Chess, we hope to make the opening
play quite sophisticated. We plan to implement the Fast
Fourier Transform in such a way that the ply-one preprocessor
will use a Fast Fourier as applied to all the Tal or Fischer
games, and hence order its moves, not based on simple child-like
heuristics, but based on the Fast Fourier of the thousands of
Tal and Fischer games. The idea would be to give the program
a chess style by having it duplicate the opening play of a
great master and then mimic the middle and end-game play of
the master once the opening book has been exhausted, backed
up with tactical brute-force search of course. To the best of
our knowledge, this idea is unique. You could help by typing
in all the great games of a Master, or if you know of such
a collection on-line, you could help by telling us about it.
Currently, we have hundreds of games by former world champion
Mikhail Tal (stored in "bookin.tal" in the distribution).
Many of these games involve lovely sacrifices and are 
quite interesting to play through.

To teach GNU Chess the games of a Master, simply get a chess
book, run gnuchess, type "neither", and then type in the game,
move by move. You may have to convert the syntax of the published game
(English or Algebraic) to gnuchess syntax (e.g. e2e4, etc.)
After you have typed the game in, use the "enter" command
to enter it in the opening book. You will then need to
use the "book" command to have these changes installed
in the binary database itself.

When you do this, GNU Chess will try to "mail" each new
game you enter to the GNU Chess opening book maintainers.
If you are not on our network, then obviously the mail
will fail -- in this instance, you should send the new
entries to us via postal mail (our postal address is
listed at the end of this document). Please do this so
that we can have your contributions.

Tree Search
-----------
GNU Chess tree search is patterned after the ply-1 preprocessor
search invented by Gillogly and Berliner. This style of search
is not too difficult to implement and carries with it an added bonus.
The search speed is not significantly interfered with by the addition
of chess knowledge.

A ply-1 preprocessor search is simply a 1-ply search, with full
evaluation taking place at ply-1 nodes, sorting the resultant
backed-up values (no need for minimax/alphabeta at this point).
The resulting sorted list is then searched, from most valuable
move to least, with a full-width depth-first alpha-beta minimax search.

A pre-set terminal depth is reached at which the depth-first search
either terminates returning its value to the previous level or
continues with a quiescence search to another pre-set level.
The quiescence search involves the investigation only of captures
or checks. In the case of this program, only captures are investigated
beyond DEPTH nodes (defined in Makefile). These captures are then
searched until MAXQUIES nodes (also defined in Makefile). 

Terminal nodes are considered to be either a) MAXQUIES nodes 
or b) DEPTH nodes if no captures exist. In either case, the
evaluation consists solely of the difference of material between
White and Black. This evaluation is then backed up in a minimax
fashion using alpha-beta vertically through the tree. At the
root node, a value and best move are returned. These are then
selected as the program's move.

Additionally, a technique called "aspiration search" is used.  This
reduces the size of the alpha-beta search window to only two pawns
wide before conducting the search. This results in more cutoffs and a
faster search (3% to 40% faster) than pure alpha-beta. However,
occasionally the returned search value will be outside of the window
and the search is then called "failing high" or "failing low". If this
happens, a second search must be done with a wider window. This may
result in an occasional slower search, but the overall improvement
averaged over many moves is such that "aspiration search" makes the
program somewhat faster.

So, essentially, GNU Chess orders the top level moves from
which it must select using any set of heuristics we devise. Then,
the search merely confirms that the move we choose wins the
most amount of material, or at least remains equal. A tiebreak
of equally valuable moves is based upon the order in which the
moves appear in our heuristically-sorted movelist at the first
ply.

Incremental search is not particularly useful with this scheme.

Parallel Processors
-------------------
GNU Chess can use more than one processor. The PARALLEL flag
in Makefile creates a GNU Chess which uses parallelism.
However, at present, the processors must share the same disk
and communicate over a network using the Berkeley
'rsh/rlogin/etc' convention. One such configuration is a SUN
network with a central fileserver.  The distributed version
of GNU Chess was set up to work with our SUN network
configuration.

This configuration is described in the 'procarray[]' in module
parallel.c. Each host is listed with its corresponding machine
architecture, whether to use it, and if it shares the same disk.
The same disk is assumed to mean the one on which the user
originally invoked GNU Chess intending to eventually use
parallelism.

Parallelism introduces many "sticky" factors into chess computing.
One of these factors is the communications overhead in managing
multiple processors. In our implementation, this overhead is
immense and prevents the search from being nearly as fast as one
would expect. But then again, our implementation is quite young
and has much work ahead of it.

Assuming N processors, the ply-1 move list is divided equally
among the N processors and each processor then searches its own
subset of the tree using the normal search algorithm, slightly
changed from the version in alphabeta.c and called 'psearch'
residing in parallel.c.

A parallel search in GNU Chess usually consumes more terminal
nodes, because good alpha and beta values are not as readily
available. Some processors may get good alpha and beta values,
others may not. Since these latter instances will slow down the
search, the entire parallelism is as strong as its weakest link.
In this case, the search terminates when the last processor 
finishes. Needless to say, many improvements can (and will) be
made to this scheme.

We also plan to introduce modifications to allow the use of
processors that don't share the same disk. This will involve
transferring the statistics of the search to the originating
processor after completion by any given processor.

Fancy Displays
--------------
Currently, GNU Chess supports two different styles of windowing
support for a fancy board. Fancy boards make it easier to play
the game, because all the chess-pieces are displayed in lovely
chess fonts on a big bitmap screen. How this generally happens
is as follows: GNU Chess is invoked through the windowing
support program which communicates with GNU Chess as a separate
process, talking with GNU Chess in order to transmit and receive
moves. All communication with GNU Chess, when using a windowing
support program, is done through the windowing support system,
generally via the use of a "mouse".

The first style of chess windowing support is for the X windowing
system. We provide full sources to the X-chess windowing display
manager within the sub-directory 'Xchess' of this distribution.
X-chess has quite a few more features than SUN's chesstool. For
this reason, the truly supported window system for GNU Chess is
X-chess. If you find a bug in X-chess, you may report it to us. 

The second style of chess windowing support is SUN's chesstool.
It is fully documented in SUN's manual pages. If you find a bug
in chesstool, you may report it to us, but you should also
report it to SUN so that they can fix it.

Conclusion
----------

GNU Chess has been semi-debugged as of 1986, seems to play on
a wide variety of machines, but nothing is guaranteed. You can
help this program by making it better.
 
The file CHANGES contains a complete history of changes made to the
program, ordered chronologically.

The file TODO contains other suggestions for how to improve this program.

Here is a list of the source files and their actual purpose:

	alphabeta.c	- searches the tree
	book.c		- implements opening book.
	eval.c		- evaluates a position
	gen.c		- generates legal moves
	hash.c		- tries to do hashing of positions already seen
	main.c		- accepts commands, executes commands
	moves.c		- random move routines
	parallel.c      - parallel processing routines
	pps.c		- positional pre-sort
	tbl_gen.c	- table-driven move generation controller
	tbl_genpc.c     - table-driven move generation for pieces
	tbl_genpwn.c    - table-driven move generation for pawns
	util.c		- random utilities
        Bitmapper/*     - complex bitmap manipulation routines
	Xchess/*        - the X window interface to GNU Chess
	bookin		- human-readable book in GNU Chess move format,
			  parsed via "book" command
	bookin.tal      - collected games of Mikhail Tal,
			  parsed via "enter" command
	bookin.bdg	- additional book games in reduced algebraic format
	convert.el      - LISP macro (in GNU Emacs Lisp) to convert
                          Xchess-format games (via 'xchess -i') to
                          GNU Chess format.

Please note that the following are not considered debugged:
table-driven move generation and hashing of positions during
tree-search.

The Bitmapper code is an experimental idea that would
eventually replace the regular code. It is completely
different in terms of the way it implements data structures
and static evaluation, the idea being that by making the
data structures more complex and informative, the program
could carry along a lot more information from move to move.
The ideas here are loosely based on the work of some Soviets
and some early researchers at Northwestern University. Note
that much of this code is hard-wired to specific machine
word sizes, in this case, a 32-bit word. But it should not
be too hard to change this to work on any given machine,
as long as the word-size perfectly divides 64.

Please remember that GNU Chess is purely experimental and
has not participated in any rated tournaments.  You are
welcome to enter it in a tournament, but proper
recognition must be given to the author and the Free
Software Foundation.  If any spectators inquire about the
program during the tournament, you are also morally bound
to tell them how you got the program and either to supply
them with the distribution you received from us, or for
the latest version tell the spectator our address so that
he may get it directly from us.  We would also like you to
tell us if you are entering it in a tournament.

Only through your contributions to GNU Chess will it
become a stronger program. There is the possibility that
it could become the strongest chess program by the end of
the 20th century if enough people join in and contribute!
Both hardware experts and software experts are needed.

If you have a hardware implementation of some of the functions
normally carried out in software, and if you would like to interface
your special-purpose hardware to GNU Chess, we will be more than happy
to talk with you.

If you have large chess databases, either of patterns, chess
knowledge, or game-listings, and if you feel these could be
of use in the development of GNU Chess, we want to talk with you.

If you make any changes, additions, modifications, improvements,
please be sure to pass along all your modifications to the address
listed below.

We will provide a tape along with GNU Chess, its full and most recent
sources, as well as the opening book(s), but there is a $150 handling
charge if you want to get GNU Chess from us directly. If you are on
our network, we can provide it to you free of charge because the
distribution doesn't take any of our time.  If you get GNU Chess from
someone else, obviously they cannot charge you, but you also won't
necessarily be getting the latest copy.

   Stuart Cracraft
   P.O. Box 13123
   Torrance, Ca. 90503

The predecessors of GNU Chess
-----------------------------

GNU Chess by Stuart Cracraft is a descendent of the Technology Chess
Program by Jim Gillogly. A history of the latter is in order. The
original TECH was written years ago by Jim when he was at
Carnegie-Mellon University working for his Ph.D. under Hans Berliner.
The results of his effort was a program called TECH, written in BLISS,
a high-level systems programming language running under the TOPS-10
operating system environment. (Jim provided some helpful theoretical
commentary during the writing of GNU Chess.)

TECH did not play very well. By most standards it was a low class D
player. TECH's chief contribution to computer chess theory was it
showed very little knowledge was necessary in order to play passable
chess.  Indeed, TECH used only material evaluation at terminal nodes
in the search tree.  The only sophisticated chess knowledge was used
at the very top of the tree to examine the positions after each of the
possible chess moves for the side on move, usually a number no larger
than fifty.

The advantage of this scheme was that huge amounts of chess knowledge
could be embedded in the program without degrading overall
performance.  However, one point in Jim's thesis was that the TECH
program eschewed this knowledge but did not forbid it.  The bent of
the program, however, was definitely to avoid it.  If the knowledge
were made a significant part of the tree-search at the top-level and
because this knowledge was applied in only 50 or so positions, and
would not be subject to the exponential tree-growth, as in the case

CHANGES
			  GNU CHESS HISTORY
		      (current version 1.2.87)


January 2, 1987   -- Stuart Cracraft at UCLA
		     He added a few more Tal games to the collection.

January 2, 1987   -- Jim Aspnes at MIT
		     He contributed MCO variations for the Catalan,
		     Queen's Indian, and Reti openings.

December 29, 1986 -- Jim Aspnes at MIT
		     He contributed all MCO variations of the Najdorf
		     to the opening book.He also contributed a LISP
		     macro (written in GNU Emacs Lisp) to convert
		     xchess game formats to GNU Chess opening book
		     format. 

December 14, 1986 -- Ken Thompson at Bell Labs
		     He contributed almost 200 games by Tal to
		     our collection of Tal-games, bringing the
		     total number of Tal positions in the book
		     to 10,692. Total book positions now 13,207.
		     These reside in bookin, bookin.bdg, bookin.tal.
		     Note that presently, only bookin and bookin.tal
		     can be used. The new Tal positions came in a
		     slightly different format, which we have chosen
		     to adopt as our standard format. All book
		     games in bookin and bookin.bdg will gradually
		     change into the new standard format.

December 11, 1986 -- Stuart Cracraft at UCLA
		     Added "averages" for node-count per move,
		     cpu per move, rate per move to list_history
		     and write_history.
		     New version of Xchess installed.
		     Started typing in Tal games into "bookin.tal".
		     Added "total book positions" printout to "book"
		     and "enter" statistics printout.

December 10, 1986 -- Stuart Cracraft at UCLA
		     Implemented aspiration search in normal
		     alpha-beta search. Speedups of 3% to 40%
		     have been noticed in most positions.
		     Occasionally a slower search will result,
		     but it is thought these are worth the
		     usual speedups.

December 9, 1986  -- Stuart Cracraft at UCLA
		     Fixed minor bug in write_history()
		     Added another Tal game, 2nd game of 1st world
		     championship match with Botvinnik, a Benoni.

December 9, 1986  -- Stuart Cracraft at UCLA
		     Enhanced parallelism. All parallel processors
		     now communicate via a shared data file and
		     are kept running (in idle loops watching the
		     shared data file). This saves us a few seconds
		     on each move since the 'rsh' need not be invoked
		     more than once (at the beginning). Since the
		     shared data file is now implemented, we will
		     next work towards a "parallel anarchy" in which
		     any processor can use any other processor in
		     order to reduce its search. The current scheme
		     with the program being only as fast as its slowest
		     processor, is quite inefficient.

December 1, 1986 --  ASP at MIT Athena
		     Added a couple of Master games from 
		     Modern Chess Openings 12 (a Fischer game,
		     and a Matanovic game).

November 30, 1986 -- Stuart Cracraft at UCLA
		     Added parallelism. Can now handle multiple
		     processors (sharing same disk). Later we will
		     add the capability to use processors not sharing
		     the same disk. Modified README and MAN-PAGE.

November 26, 1986 -- Stuart Cracraft at UCLA
		     Fixed a few bugs in book-mailing mechanism.
		     Fixed a bug regarding situations where only
		     one move is available.
		     Fixed a bug in read_history() that caused
		     Black queenside castles to be mishandled.

November 25, 1986 -- Stuart Cracraft at UCLA
		     Added two pawn heuristics. Reward pawns moving into
		     a phalanx of pawns. A phalanx is two or more
		     horizontally-connected pawns. Likewise, penalize
		     pawns leaving a phalanx of pawns. The penalty for
		     leaving is a little more than the reward for
		     entering.

November 24, 1986 -- Stuart Cracraft at UCLA
		     A user reported an unbelievable bug. Investigation
		     of this bug led to the discovery that GNU Chess was
 		     not picking the move judged best by the tree search
		     in all cases. This resulted in the bug showing
 		     itself which further showed that the program was
		     selecting an inferior move. This may result in an
		     improvement to the program's play.

November 24, 1986 -- Stuart Cracraft at UCLA
		     Added two heuristics. Penalize king moves if
		     the king hasn't castled yet. Also, penalize pawn
		     moves which produce doubled pawns. Should
		     probably have something for isolated pawns
		     too.

November 23, 1986 -- Wayne Christopher at Berkeley
		     New version of X chess display front-end.
		     Fixed bugs include multiple pieces, runs
		     on SUNS & Bobcats, loads saved games.

November 23, 1986 -- Stuart Cracraft at UCLA
		     Cleaned up some minor bugs regarding history.
		     Added "Illegal command" error message at Wayne's
		     request.

November 22, 1986 -- David Goldberg at SUN Microsystems
		     He complained that GNU Chess was memory-hungry.
		     A few minor modifications to hash.c reduced
		     uninitialized data space 87% and text space
		     12%. This should make it easier for GNU Chess
		     to run on small computers.

November 22, 1986 -- Stuart Cracraft at UCLA
		     "read" command was working, but needed
		     additional tweaking so that history
		     array would be printed by list_history().

November 19, 1986 -- Stuart Cracraft at UCLA
		     Added "read" command which reads a history
		     file (game listing) and restores the board
		     to as if the person was still playing that.
		     particular game. Generally cleaned up 
		     history mechanism, made it more orthogonal.
		     Revised README. Added doc to MAN-PAGE.

November 16, 1986 -- Stuart Cracraft at UCLA
		     More opening book bugs found and fixed.
		     Added capability to accept abbreviated-algebraic notation
		     for entering "book" games from files.
		     Added approximately 2500 new positions to
		     opening book from games involving the
		     opening called Blackmar-Diemer Gambit,

pps.c
/* This file contains the positional pre-sort for CHESS.
   Copyright (C) 1986 Free Software Foundation, Inc.

This file is part of CHESS.

CHESS is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.  Refer to the CHESS General Public
License for full details.

Everyone is granted permission to copy, modify and redistribute
CHESS, but only under the conditions described in the
CHESS General Public License.   A copy of this license is
supposed to have been given to you along with CHESS so you
can know your rights and responsibilities.  It should be in a
file named COPYING.  Among other things, the copyright notice
and this notice must be preserved on all copies.  */

/*
 * Please note: This is where all the important chess knowledge lives.
 * If you add code here, your code does not have to be efficient.
 * Because the knowledge is applied to (at most) 50 or so positions,
 * this happens in a fraction of a second. You can add as much
 * knowledge as inefficiently (or efficiently!) as you like.
 * The search will not be slowed down by the addition of new chess knowledge.
 */

#include <stdio.h>
#include "gnuchess.h"

#define PK2K4       30		/* Bonus for pawn to K4 from K2*/
#define PK3K4        2		/* Bonus for pawn to K4 from K3 */
#define PQ2Q4       20		/* Bonus for pawn to Q4 from Q2 */
#define PQ3Q4        2		/* Bonus for pawn to Q4 from Q3 */
#define KCAS       100		/* Bonus for king-side castle */
#define QCAS        75		/* Bonus for queen-side castle */
#define NR3        -15		/* Penalty for knight moves to rook-3  */
#define BLOKPWN    -50		/* Penalty for blocking center pawn w/ piece */
#define LIBBISH     30		/* Bonus for liberating undeveloped bishops */
#define PCMOVK       2		/* Bonus for moving piece on king side */
#define CAPCENT      5		/* Bonus for pawn captures toward center */
#define CAPACENT    -5		/* Penalty for pawn caps away from center */
#define CAPISOP    -10		/* Penalty for pawn cap giving isolated pawn */
#define WNGPADV    -20		/* Penalty for advancing pawns on the wing */
#define CAPCENTUSUP 50		/* Bonus for taking an unsupported K/Q pawn */
#define CAPCENTSUP -15		/* Penalty for capturing supported K/Q pawn */
#define KMOVE      -40          /* Penalty for moving a king before castling */
#define PDOUB      -15		/* Penalty for pawn move which creates a
				   doubled pawn */
#define PFLANKE      5          /* Bonus for pawn entering phalanx of pawns */
#define PFLANKL     -7          /* Penalty for pawn leaving phalanx */
/* add these for opening before 10th move */
#define N2NDBD       5		/* Knight develop to 2nd rank after
				   corresponding bishop developed */
#define NBISH3      15		/* Knight develop bishop 3 */
#define EQUEENM	   -40    	/* Early queen move penalty */
        			/* Decreased by 10 for each minor piece
				   that has ever developed */
/* and these */
#define ROOKOPEN    40		/* Rook on pure open file (no pawns) */
#define ROOKSOPEN   25		/* Rook on semi-open file (attacking 
				   unsupported enemy pawn) */
#define ROOKSEVEN  100		/* Rook to seventh rank */

int     sqcontrol[120] = {	/* Value of squares on board */
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 1, 2, 3, 3, 2, 1, 0, 0,
    0, 1, 3, 4, 5, 5, 4, 3, 1, 0,
    0, 2, 4, 6, 7, 7, 6, 4, 2, 0,
    0, 3, 5, 7, 8, 8, 7, 5, 3, 0,
    0, 3, 5, 7, 8, 8, 7, 5, 3, 0,
    0, 2, 4, 6, 7, 7, 6, 4, 2, 0,
    0, 1, 3, 4, 5, 5, 4, 3, 1, 0,
    0, 0, 1, 2, 3, 3, 2, 1, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
    0, 0, 0, 0, 0, 0, 0, 0, 0, 0
};

/*
 * Multiplier of sqcontrol based on piece-type, e.g. bonus for moving to square */
int cent[MAXPC+1] = {0, 1, 4, 3, 2, 1, -1}; 
extern int pcval[MAXPC+1];
extern histtot,sorttype,propiece,compare();

/* Return castling bonuses if appropriate */
ppscastle(movpiece,flags)
char flags;
{
    if (abs(movpiece) == WK)
      {
	if (flags&KCASFLAG) return(KCAS);
        else if (flags&QCASFLAG) return(QCAS);
      }
}

/* Modify centrality bonus if in middle-game */
ppsmiddle(from,to)
{
    cent[1] = 3;
    cent[2] = 4;
    cent[3] = 3;
    cent[4] = 2;
    cent[5] = 1;
    cent[6] = 1;
    return(0);
}

/* Modify centrality bonus if in end-game */
ppsending(from,to)
{
    cent[1] = 0;
    cent[2] = 4;
    cent[3] = 3;
    cent[4] = 1;
    cent[5] = 1;
    cent[6] = 4;
    return(0);
}

/*
 * Opening heuristics
 */
ppsopening(bd,from,to,movpiece,flags)
struct bdtype bd[120];
char flags;     
{
        int nundminors;
	int posval = 0;
	switch (movpiece) {
	    case WN:
	    case BN:
		if ((movpiece == WN && from == 92 && bd[93].piece == WB) ||
		    (movpiece == WN && from == 97 && bd[96].piece == WB) ||
		    (movpiece == BN && from == 22 && bd[23].piece == BB) ||
		    (movpiece == BN && from == 27 && bd[26].piece == BB))
		    posval += N2NDBD;
		break;
	    case WB:
	    case BB:
		if (((from == 92 && to == 73) || (from == 97 && to == 76)) ||
		    ((from == 22 && to == 43) || (from == 27 && to == 46)))
		    posval += NBISH3;
	        break;
	    case WQ:
	    case BQ:
		if (histtot <= 10)
		{
		    if (movpiece == WQ)
			nundminors = (bd[92].piece != WN +
				      bd[93].piece != WB +
				      bd[96].piece != WB +
				      bd[97].piece != WN);
		    else
			nundminors = (bd[22].piece != BN +
				      bd[23].piece != BB +
				      bd[26].piece != BB +
				      bd[27].piece != BN);
		    /* Decrease queen penalty by 10 for each developed
		       minor piece */
		    posval += (EQUEENM + (nundminors * 10));
		}
	}
    return(posval);
}	

ppsheuristics(bd,from,to,movpiece,flags)
struct bdtype bd[120];
char flags;
{
    int i,doub;
    int posval = 0;
    posval -= (bd[from].moved*5);	/* Penalize early multiple moves */
    switch(movpiece) {
       case WP: 
       case BP:
	        posval += (((to % 10 > 1 && bd[to-1].piece == movpiece) ||
			   (to % 10 < 8 && bd[to+1].piece == movpiece)) ?
			   PFLANKE : 0) +
			  (((from % 10 > 1 && bd[from-1].piece == movpiece)
			    || (from % 10 < 8 && bd[from+1].piece == movpiece))
			   ? PFLANKL : 0);
	    	posval +=
	                  (((movpiece == WP) && (from % 10 <= 4) &&
			    ((from - to) % 9 == 0)) ||
			   (movpiece == BP && (from % 10 <= 4) &&
			    ((from - to) % 11 == 0)) ||
			   (movpiece == WP && (from % 10 >= 5) &&
			    ((from - to) % 11 == 0)) ||
			   (movpiece == BP && (from % 10 >= 5) &&
			    ((from - to) % 9 == 0)) ? CAPCENT : 0);
		posval += 
	                  (((movpiece == WP) && (from % 10 <= 4) &&
			    ((from - to) % 11 == 0)) ||
			   (movpiece == BP && (from % 10 <= 4) &&
			    ((from - to) % 9 == 0)) ||
			   (movpiece == WP && (from % 10 >= 5) &&
			    ((from - to) % 9 == 0)) ||
			   (movpiece == BP && (from % 10 >= 5) &&
			    ((from - to) % 11 == 0)) ? CAPACENT : 0);
		posval +=   (((movpiece == BP && ((from == 34 && 
			      bd[23].piece == BB) || ((from == 35 &&
			      bd[26].piece == BB)))) ||
			   ((movpiece == WP && ((from == 84 && 
			      bd[93].piece == WB) || ((from == 85 &&
					       bd[96].piece == WB))))))
			    ? LIBBISH : 0);
		posval +=
	                  (((from == 85 && to == 65) ||
			    (from == 35 && to == 55)) ? PK2K4 : 0) +
			  (((from == 75 && to == 65) ||
			    (from == 45 && to == 55)) ? PK3K4 : 0);
		posval +=
			  (((from == 84 && to == 64) ||
			    (from == 34 && to == 54)) ? PQ2Q4 : 0) +
			  (((from == 74 && to == 64) ||
			    (from == 44 && to == 54)) ? PQ3Q4 : 0) +
			  (((from % 10 == 1 || from % 10 == 8) &&
			    (to % 10 == 1 || to % 10 == 8)) ? WNGPADV : 0);
		doub = 0;
		for (i = from; i < 99; i+=10)
		    if (bd[i].piece == movpiece)
			doub++;
		for (i = from; i > 20; i-=10)
		    if (bd[i].piece == movpiece)
			doub++;
		posval += (doub != 0 ? PDOUB : 0);
		break;
      case WN:
      case BN: posval += ((movpiece == WN && (to == 71 || to == 78))
			       ? NR3 : 0) +
			     ((movpiece == BN && (to == 31 || to == 38))
			       ? NR3 : 0) +
				 ((from % 10 >= 5) ? PCMOVK : 0);
		posval += (((movpiece == WN) && ((to == 74 && 
						bd[84].piece == WP) ||
			   (to == 75 && bd[85].piece == WP))) ||
			 (((movpiece == BN) && ((to == 44 && 
						bd[34].piece == BP) ||
  			  (to == 45 && bd[35].piece == BP)))) ? BLOKPWN : 0);
                break;
      case WQ:
      case BQ:	posval += (((movpiece == WQ) && ((to == 74 && 
						bd[84].piece == WP) ||
			   (to == 75 && bd[85].piece == WP))) ||
			 (((movpiece == BQ) && ((to == 44 && 
						bd[34].piece == BP) ||
  			  (to == 45 && bd[35].piece == BP)))) ? BLOKPWN : 0);
		break;		
      case WK:
      case BK:
		posval += ((bd[(movpiece == WK) ? 93 : 23].moved == 0) &&
			    (to != 97 && to != 93 && to != 27 &&
			     to != 23) ? KMOVE : 0);
   		break;
      case WB:
      case BB: posval += ((from % 10 == 5 || from % 10 == 6 ||
			      from % 10 == 7 || from % 10 == 8)
			       ? PCMOVK : 0) +
			 (((movpiece == WB) && ((to == 74 && 
						bd[84].piece == WP) ||
			   (to == 75 && bd[85].piece == WP))) ||
			 (((movpiece == BB) && ((to == 44 && 
						bd[34].piece == BP) ||
  			  (to == 45 && bd[35].piece == BP)))) ? BLOKPWN : 0);

                break;
      }
      return(posval);
}			    

pps(bd,moves)
struct bdtype bd[120];
struct mvlist moves[MAXMOVES];
{
    register i,from,to,posval,width,twidth,ourkloc,hiskloc,incheck,
	     real_width,spec;
    struct mvlist tempmoves[MAXMOVES];
    /* Scan board trying to find our king */
    ourkloc = locate_king(bd);
    /* Set variable incheck according to if king of side-to-move */
    /* is in check */
    incheck = sqattacked(bd,ourkloc,OPPCOLOR(bd[TOMOVE].moved));
    /* Generate all moves, pseudo-legal or otherwise */
    width = generate(bd,moves);
    real_width = width;
    for (i = 0; i < width; i++)
      {
        posval = 0;
	twidth = 0;
	/*
         * The following is necessary to prune out all moves which
	 * leave a checked king in check.
	 */
	makemove(moves[i],bd);
	bd[TOMOVE].moved = OPPCOLOR(bd[TOMOVE].moved);
	ourkloc = locate_king(bd);
	bd[TOMOVE].moved = OPPCOLOR(bd[TOMOVE].moved);
	incheck = sqattacked(bd,ourkloc,bd[TOMOVE].moved);
	unmakemove(moves[i],bd);
	if (incheck)
	  {
/*
                lin_to_alg(moves[i].from,stdout);
	        lin_to_alg(moves[i].to,stdout);
	 	printf(" leaves computer in check... discarding\n");
*/
		moves[i].score = MAXINT;
		real_width--;
		continue;
	  }
	from = moves[i].from;
	to = moves[i].to;
#ifdef SPPSDEBUG
        lin_to_alg(from,stdout);
	lin_to_alg(to,stdout);
	printf(" = ");
#endif
/*  temporarily disable this, not sure how good it is.
	if (histtot <= 10)
	  posval += ppsopening(bd,from,to,moves[i].movpiece,moves[i].flags);
*/
	/* Give castling credit and heuristic credit */
	posval += ppscastle(moves[i].movpiece,moves[i].flags) +
	          ppsheuristics(bd,from,to,moves[i].movpiece,moves[i].flags);
	/* Modify mobility credits according to phase of game */
	if (bd[WMAT].moved <= 6920 || bd[BMAT].moved <= 6920)
	  posval += ppsending(from,to);
	else
	  posval += ppsmiddle(from,to);
	/* Now credit extra mobility */
	makemove(moves[i],bd);
	twidth = 0;
	bd[TOMOVE].moved = OPPCOLOR(bd[TOMOVE].moved);
	twidth = generate(bd,tempmoves);
    	posval += twidth;
	unmakemove(moves[i],bd);
	bd[TOMOVE].moved = OPPCOLOR(bd[TOMOVE].moved);
#ifdef SPPSDEBUG
	printf("ppsphase(%d) + cent(%d) + cap(%d) + prom(%d)=",
          posval,(sqcontrol[to]-sqcontrol[from])*cent[abs(moves[i].movpiece)],
	       (moves[i].flags&CAPFLAG) ? pcval[abs(moves[i].cappiece)] : 0,
	       (moves[i].flags&PROMFLAG) ? 
	       pcval[abs(moves[i].propiece)] : 0);
#endif
	/* Credit square control */
	posval += (sqcontrol[to]-sqcontrol[from])*
	  cent[abs(moves[i].movpiece)];
	/* Force captures to top of list */
	if (moves[i].flags&CAPFLAG)
	  posval += pcval[abs(moves[i].cappiece)];
	/* Force promotions to top of list */
	if (moves[i].flags&PROMFLAG)
	  posval += pcval[abs(moves[i].propiece)];
#ifdef SPPSDEBUG
	printf("%d\n",posval);
#endif
        moves[i].score = -posval;
      }
    /* Sort the whole list using the scores, best move to worst move */
    sorttype = SCORESORT;
    qsort(moves,width,sizeof(struct mvlist),compare);
    return(real_width);
}
